### ISSUE TITLE
array.reindex inverse order

### ISSUE THREAD

**Issue by Dunky13** [chapel-lang/chapel#12106]
### Summary of Problem
I was hoping to reverse a 2D array, but the array.reverse only works on 1D arrays (https://github.com/chapel-lang/chapel/blob/master/modules/internal/ChapelArray.chpl#L3216). array.reindex was another potential solution. https://github.com/chapel-lang/chapel/blob/master/modules/internal/ChapelArray.chpl#L2574

But running `array.reindex(0..x by -1, 0..y by -1)` seems to ignore the "by -1"

### Proof of concept:
```chapel
module ReverseReIndex {
  const data_size = 8
  proc main() {
    var data = [i in 0..#data_size] i;
    writeln(data); // 0 1 2 3 4 5 6 7
    ref revData = data.reindex(0..#data_size by -1);
    writeln(revData); // 7 6 5 4 3 2 1 0
  }
}
```

**Comment by bradcray** [comment]
@Dunky13: Can you say more about what behavior you'd like to see when reversing a 2D array?  Are you imagining reversing it relative to a 1D linearization of the 2D array?  If so, what order would be used to linearize the array?  (e.g., row- or column-major order, or something else).  Apologies if there's an obvious precedent here that I'm unaware of.

**Comment by Dunky13** [comment]
Currently working on an assignment where I have to rotate a 2D array by 180deg. Like:
[[1,2,3],
[4,5,6]]
V
[[6,5,4],
[3,2,1]]

I work with 500+ 3000x4000 2D arrays. I like the reindexing as it passes a reference. Which I assume doesn't increase memory significantly. Currently using:

```chapel
proc rotate(A: [?D] ?t) {
  return reshape (A, D by -1);
}
```

Which works, but returns a new array. And since I'm reusing the non-rotated array later on, it becomes quite memory intensive.

If reindexing is just changing the reference of reading order, then it would speed up my process quite a bit.

Hope that explains the precedent a bit.

I was quite confused that reshape allows me to define 'D by -1' but reindex seems to scrap the stepping order.

**Comment by Dunky13** [comment]
As you've summed it up is exactly what I'd need yes.

Additional operations are: arr_i = non-rotated array, arr_j is rotated. Apply FFT Forward to both arrays and multiply the results. This loops over the N arrays mentioned earlier, where each arr_i has to be rotated at some point, thus the use for reference.
I'm storing the multiplication in a separate arrays with same dimensions as further calculations/operations happen on those results.

**Comment by bradcray** [comment]
Further questions:
* Is there a precedent for this type of reversed multidimensional array indexing in another language that we should look to for reference?
* Would it be sufficient to have a method you could call on the array to index into it in reversed order?  e.g., I'm imagining something like `A.access(i, j, reverse_i=true)` or `A.access(i, j, reverse_i=true, reverse_j=true)`
* Are you writing your own FFTs or using a library like FFTW?

**Comment by Dunky13** [comment]
To answer your last question first,
I'm using the FFTW library, as doing own FFT is out of assignment scope, and to be honest outside of skill set as well. So I'm thankful for that library :-)

Given that, I don't know if access reverse would work, as it needs to create a plan, based on the array and it's reference. So I thought that the reindexing function would allow for a neat and performant solution.

I am unaware of any precedent in other languages, though I wouldn't be surprised if there are solutions available in other languages. I haven't found, nor have I looked for them.

The main reason I opened the bug report is that I found it curious that the `reshape` function allows for a negative stride, but `reindex` seems to strip the (negative) stride. (Haven't tested twice the domain and `by 2` so can't say for sure in the positive stride space)

**Comment by bradcray** [comment]
@Dunky13: I think you're right that, given that it's an external library, the reversing accessor method wouldn't help (the FFTW library wouldn't know to use it). But for that matter, I also think that any method short of actually moving the elements around in memory (whether in the same array or a different one) wouldn't either.  So unless FFTW has a way of specifying that dimensions should be reversed, I think we'd have to copy stuff around.

That said, I'm guessing that FFTW does have a way to do so (i.e., I'm guessing that if you want it, others would too?), in which case maybe we've failed to expose it in the Chapel interface or have overlooked it?  Scanning the [documentation for Chapel FFTW](https://chapel-lang.org/docs/primers/FFTWlib.html) quickly, I'm seeing references to create reverse plans:  https://chapel-lang.org/docs/primers/FFTWlib.html#complex-complex-in-place-transform.  Is this the sort of thing you need?

[I'm also tagging @npamana, as the local FFTW expert... I know dangerously little about them]

Surprise about reindex vs. reverse vs. reshape noted, but I'm more interested in finding you a good solution right now than wrestling with features that I'm not sure your use case should need to rely on.

**Comment by npadmana** [comment]
@Dunky13 -- here are a first set of thoughts here, limited to the question of the FFT of a reversed array.
You might have already considered/rejected these for some reason -- so apologies if this is repeating anything.

* you can determine the FFT of the reversed array simply from the FFT of the actual array (see the code listed below).  This way, you can deal with the multiplications etc without ever explicitly computing the FFT of the reversed array. If you wanted to be fancy, you could even write a Chapel iterator/`this` method that did this for you.
* If you are storing the FFT of both the actual and reversed array, then you might just do an in-place transform. 
* FFTW does have an advanced interface that I believe allows for -ve strides. I haven't had a chance to try it and we didn't expose it in the Chapel module, but we could do that if that would be useful (http://www.fftw.org/fftw3_doc/Advanced-Complex-DFTs.html#Advanced-Complex-DFTs).

```chapel
use FFTW;
use Random;

config const n1=2;
config const n2=2;
config const verbose=true;

const D = {0.. #n1, 0.. #n2}; // Just an example
// a1 : The original array
// a2 : The reversed array
// b1 : FFT of a1
// b2 : FFT of a2
// c1 : FFT of a2, but calculated from b1
var a1,b1,a2,b2,c1: [D]complex;


fillRandom(a1); // Original array
a2 = reshape(a1, D by -1);

if verbose {
  writeln("The original array :");
  writeln(a1);
  writeln("The reversed array :");
  writeln(a2);
}

// Perform the FFT
var plan1 = plan_dft(a1, b1, FFTW_FORWARD, FFTW_ESTIMATE);
var plan2 = plan_dft(a2, b2, FFTW_FORWARD, FFTW_ESTIMATE);
execute(plan1);
execute(plan2);

// Now transform the FFT of the original array into
// the FFT of the reversed case. 
forall (i,j) in D {
  const i1 = if i >0 then n1-i else i;
  const j1 = if j >0 then n2-j else j;
  c1[i1,j1] = b1[i,j]*exp(-2i*pi*i/n1:real)*exp(-2i*pi*j/n2:real);
}

if verbose {
  writeln("The FFT of the original array, reversed :");
  writeln(c1);
  writeln("The FFT reversed array :");
  writeln(b2);
}

// Verify
var diff = 0.0;
forall (expected, actual) in zip(b2, c1) with (max reduce diff) {
  diff = abs(expected-actual);
}
writeln("Maximum difference :", diff);
```

**Comment by npadmana** [comment]
A quick follow up — all that the code is doing is mapping negative frequencies to positive frequencies (with a phase)

Also, if your array is real, then the negative and positive frequencies are complex conjugates of one another, so you can simplify the code, since you don’t need to reverse frequencies in this case.

**Comment by bradcray** [comment]
> I found my own workaround as mentioned in my earlier messages

Fair enough.  As mentioned before, the reason I keep focusing on the FFT part of the conversation is because you mentioned that your workaround is resulting in an unnecessary array copy, so I was hoping to get you to a solution that was more efficient and felt like less of a workaround.

That said, I have not lost sight of the original request being more about surprise regarding the seemingly inconsistent behavior between reshape and reindex with a negative stride.

**Comment by ben-albrecht** [comment]
In addition to the in-place array rotations requested here, I think array-copy equivalents would be generally useful as well. For example, numpy provides:

- [`numpy.flip`](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.flip.html)
- [`numpy.flipud`](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.flipud.html#numpy.flipud)
- [`numpy.fliplr`](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.fliplr.html#numpy.fliplr)
